diff --git a/node_modules/heatmap.js/build/heatmap.js b/node_modules/heatmap.js/build/heatmap.js
index 3eee39e..3737abd 100644
--- a/node_modules/heatmap.js/build/heatmap.js
+++ b/node_modules/heatmap.js/build/heatmap.js
@@ -524,7 +524,7 @@ var Canvas2dRenderer = (function Canvas2dRendererClosure() {
 
       }
 
-      img.data = imgData;
+      // img.data = imgData;
       this.ctx.putImageData(img, x, y);
 
       this._renderBoundaries = [1000, 1000, 0, 0];
